source/Stargate-Model/WebOrigin.class.st

```diff
@@ -0,0 +1,67 @@
+Class {
+	#name : #WebOrigin,
+	#superclass : #Object,
+	#instVars : [
+		'scheme',
+		'host',
+		'port'
+	],
+	#category : #'Stargate-Model-CORS'
+}
+
+{ #category : #'instance creation' }
+WebOrigin class >> basedOn: aUrl [
+
+	( self hasValidOrigin: aUrl )
+		ifFalse: [ InstanceCreationFailed signal: ( '<1p> does not comply with a valid origin' expandMacrosWith: aUrl ) ].
+
+	^ self new initializeBasedOn: aUrl
+]
+
+{ #category : #preconditions }
+WebOrigin class >> hasValidOrigin: aZnUrl [
+	" According RFC6454, the Origin should have the format -- scheme '://' host [ ':' port ] --
+   with <scheme>, <host>, <port> from RFC 3986.
+	The // precedes the authority component of the uri (RFC3986), and thus we filter the schemes
+	that may have the authority component. We may add any valid scheme as we want. "
+
+	| validOriginSchemes |
+
+	validOriginSchemes := #(#http #https).
```

Comment:
I don't like the idea of asking if it's a valid object once already exists. Is it possible to move this assert to the instance creation message?

Reply:
it is on the class side

Reply:
Sorry bo'